src/cpp/shared/game/game.cpp

```diff
@@ -272,6 +273,75 @@ void philAI::RedistributeTroops(armyGroup *army1, armyGroup *army2, int a1, int
 	}
 }
 
+void game::InitRandomArtifacts() {
+  memset(artifactGeneratedRandomly, 0, sizeof(artifactGeneratedRandomly));
+  for (int i = 0; i < MAP_WIDTH; ++i) {
+    for (int j = 0; j < MAP_HEIGHT; ++j) {
+      const auto &tile = map.tiles[j * MAP_WIDTH + i];
+      if (tile.objType == (TILE_HAS_EVENT | LOCATION_ARTIFACT)) {
+        // Don't know why we need to divide by 2, but that's what the decompiled code says.
+        const int artifactId = static_cast<unsigned char>(tile.objectIndex) / 2;
```

Comment:
It's because each artifact has two tiles. tile.objectIndex / 2 converts the tile ID to the represented artifact ID.

Reply:
I've run across other code that refers to the low tile id and the high tile id.  Do you know/can you explain how that works?

Reply:
Not sure what you're referring to. Example?

Reply:
Line 28515 of HEROES2W.c, random artifacts get converted into two objects via game::ConvertObject.

Reply:
What you're seeing: It's choosing the artifact ID, replacing the random artifact's main tile with the main tile of the chosen artifact, and doing similar for the tile directly to the left.